Code submission for FLIP fest 'Playground Feature: Improve the resource explorer 29 - Milestone 3'

hichana opened the pull request:
Closes: N/A

## Description

This PR contains code changes for my submission for the 3rd/final milestone on [FLIP fest issue 29](https://github.com/onflow/flip-fest/issues/29).  It contains an implementation of a UI that will allow users to have a better experience with the resources explorer. I made a comprehensive guide with notes and instructions on how to run the updated UI locally [HERE](https://github.com/hichana/flip-fest/blob/submissions/issue-%2329/milestone-3/hichana/submissions/issue-29/milestone-3/hichana/README.md). 
______

For contributor use:

- [ x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/onflow/flow-playground/blob/master/CONTRIBUTING.md))
- [ n/a] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x ] Code follows the [standards mentioned here](https://github.com/onflow/flow-playground/blob/master/CONTRIBUTING.md).
- [ n/a] Updated relevant documentation 
- [n/a ] Re-reviewed `Files changed` in the Github PR explorer
- [n/a ] Added appropriate labels

MaxStalker commented on src/api/apollo/generated/graphql.tsx:
```diff
@@ -10,157 +10,43 @@ export type Scalars = {
-  UUID: any;
   Version: any;
   Address: any;
+  UUID: any;
```
Do we need to move it down? :)

hichana replied:
Thanks for reviewing all of this Max. 

In 'graphql.tsx', any changes you see are the result of me having run `npm run graphql:codegen`. The tool reorganizes the generated code. I don't believe I actually needed to run the tool tho, so if you prefer I can revert changes to this file. Or, I can just leave it as-is knowing that the tool is reorganizing the code. Please let me know which is preferrable.

MaxStalker commented on src/components/AccountList.tsx:
```diff
@@ -54,20 +56,21 @@ const AccountList: React.FC = () => {
-
                 {isActive && <ExportButton id={account.id} typeName={typeName}/>}
               </AccountCard>
+              <ResourcesExplorerButton addr={accountAddress} />
```
Can we use full `address` for a prop name? :)

hichana replied:
Changed in [this commit](https://github.com/onflow/flow-playground/pull/175/commits/289b5d04ba186213347bb27b87021fdfc8efaf3e)

MaxStalker commented on src/components/Sidebar.tsx:
```diff
@@ -34,39 +35,39 @@ const Sidebar: React.FC = () => {
         }
         onSelect={(_, id) => {
-          navigate(`/${projectPath}?type=tx&id=${id}`)
+          navigate(`/${projectPath}?type=tx&id=${id}&storage=${selectedResourceAccount || 'none'}`)
```
We can calculate and store value of `selectedResourceAccount || 'none'` in a single place to DRY it a bit.

hichana replied:
Changed in [this commit](https://github.com/onflow/flow-playground/pull/175/commits/289b5d04ba186213347bb27b87021fdfc8efaf3e)

MaxStalker commented on src/components/TemplatePopup.tsx:
```diff
@@ -0,0 +1,220 @@
+                      ), 
+                      templateName
+                    )
+                    navigate(`/${projectPath}?type=tx&id=${res.data?.createTransactionTemplate?.id}&storage=${selectedResourceAccount || 'none'}`)
```
Same DRY for `${selectedResourceAccount || 'none'}`

hichana replied:
Changed in [this commit](https://github.com/onflow/flow-playground/pull/175/commits/289b5d04ba186213347bb27b87021fdfc8efaf3e)

MaxStalker commented on src/layout/BottomBarItemInsert.tsx:
```diff
@@ -0,0 +1,23 @@
+  &:active {
+    cursor: ${p => (p.grab ? "grabbing" : "pointer")};
+  }
+`;
```
Can we add new line at eof?

hichana replied:
Changed in [this commit](https://github.com/onflow/flow-playground/pull/175/commits/289b5d04ba186213347bb27b87021fdfc8efaf3e)

MaxStalker commented on src/providers/Project/projectDefault.ts:
```diff
@@ -165,7 +165,7 @@ export function createLocalProject(
       deployedCode: "",
       deployedContracts: [],
-      state: ""
+      state: "{}"
```
Why do we need a empty object string here?

hichana replied:
I did this as a matter of convenience. For example, I use 'Object.keys' to iterate over the state object and generate the the storage items based on the result of Object.keys. By having an empty object I get zero results and if the user has no items in storage nothing is rendered. However, just as well I can set to this an empty string at the expense of adding more logic/checks when generating the storage items. Seemed like the most elegant solution was to assume this will always be an object (empty at first). 

Please let me know your preference here. If an empty string is the right way to go I can refactor how this is all handled.

MaxStalker replied:
Sure, you can use empty object to iterate with `Object.keys`, but you have a `string`, representing empty object here :)
or you `JSON.parse` it somewhere down the line?

hichana replied:
Ok, marking this one resolved after chatting with Max. In this case, the string that represents an empty object here is indeed used by JSON.parse down the line, and by doing so we are also removing the need for various other conditional checks elsewhere in the code.

MaxStalker commented on src/util/accounts.ts:
```diff
@@ -0,0 +1,8 @@
+  "0x03": 2,
+  "0x04": 3,
+  "0x05": 4
+}
```
Add new line at eof

hichana replied:
Changed in [this commit](https://github.com/onflow/flow-playground/pull/175/commits/289b5d04ba186213347bb27b87021fdfc8efaf3e)

MaxStalker reviewed (changes requested):
Looks good :)
Just some small changes here and there.

hichana commented:
@MaxStalker I made changes based on your feedback and merged to this branch/PR. I left responses to two of your comments in this PR and will wait for your response to see if more changes are needed, which I'm happy to do if so. Thank you for reviewing my submission!

MaxStalker commented on src/util/storage.ts:
```diff
@@ -0,0 +1,65 @@
+
+  return { storage, paths, types, capabilities }
+ 
+}
```
New line on eof 😛

MaxStalker commented:
@hichana can you also check that imports are using absolute paths? i.e.
```js
import { SideBar } from "components/Sidebar"
```
and not
```js
import { SideBar } from "../../components/Sidebar"
```
I've seen the later for `components`, `hooks` and `util` folders :)
Thanks in advance!

hichana commented:
@MaxStalker I added the `util` alias to both the 'tsconfig.json' file and 'webpack.config.js', which then allows the util to be used as an absolute import. I changed my use of relative imports as well as the ones that existed from previous master branch. For the `hooks`, I'm having trouble getting that to work for some reason. Actually, I don't believe I made any new imports for hooks in my submission, but I'm happy to change the existing ones if you can help troubleshoot why adding the alias for `hooks` to the two config files doesn't make the absolute imports work. And finally, can you tell me where you found `import { SideBar } from "../../components/Sidebar"`? I went through everything file by file and can't find an instance where I did a relative import for a component. Maybe I just missed it tho.

For now, I pushed my changes to the `util` absolute imports on my submission branch for this PR.

MaxStalker commented:
Looks great!
Thank you for updating the imports ;)

kerrywei commented:
@hichana if you can do a rebase, it might enable you yo merge in this PR as the CODEOWNER file has been updated (so that Max is also one of the code owners)

hichana commented:
Hi @kerrywei -- I believe I did the rebase. Is it what was needed?

kerrywei commented:
@hichana it should be good to go now
